src/jq_nif.erl

```diff
@@ -16,6 +18,41 @@
 process_json(_, _) ->
     not_loaded(?LINE).
 
+create_jq_resource() ->
+    not_loaded(?LINE).
+
+cancel_jq_resource(_JQResource) ->
+    not_loaded(?LINE).
+
+process_json_with_jq_resource(_FilterProgram, _JSONText, _JQResource) ->
+    not_loaded(?LINE).
+
+
+timeout_resource(JQResource) ->
+    case cancel_jq_resource(JQResource) of
+        ok -> ok;
+        retry ->
+            % Attempt to give dirty schedulers some time before continuing
+            timer:sleep(1),
+            timeout_resource(JQResource)
+    end.
+
+process_json(FilterProgram, JSONText, TimeoutMs) ->
+    JQResource  = create_jq_resource(),
+    {ok, TRef} = timer:apply_after(TimeoutMs, jq_nif, timeout_resource, [JQResource]),
+    Res = process_json_with_jq_resource(FilterProgram, JSONText, JQResource),
+    timer:cancel(TRef),
+    case Res of
+        {error, {timeout, _}} ->
+            ErrorMsgFormatTO = 
+                "jq program canceled as it took too long time to execute (timeout set to ~w ms)",
+            ErrorMsgTO =
+                erlang:iolist_to_binary(io_lib:format(ErrorMsgFormatTO, [TimeoutMs])),
+            logger:error(ErrorMsgTO),
+            {error, {timeout, ErrorMsgTO}};
+        Term -> Term
+    end.
+
```

Comment:
maybe also add a test case to cancel the jq resource twice?

Reply:
The function [cancel_jq_resource_nif](https://github.com/emqx/jq/blob/caf9c93bff9997b195e66aca7bf75f694f25cfa9/c_src/erlang_jq_nif.c#L302) will not be called again with the same NIF resource after it has returned ok. I don't see why we should test something that never should happen. Do you mean something else?